spek-runtime/src/commonMain/kotlin/org/spekframework/spek2/runtime/SpekRuntime.kt

```diff
@@ -9,24 +13,54 @@ import org.spekframework.spek2.runtime.execution.ExecutionRequest
 import org.spekframework.spek2.runtime.lifecycle.LifecycleManager
 import org.spekframework.spek2.runtime.scope.*
 import org.spekframework.spek2.runtime.util.ClassUtil
+import kotlin.time.ExperimentalTime
+import kotlin.time.measureTime
 
 class SpekRuntime {
-    fun discover(discoveryRequest: DiscoveryRequest): DiscoveryResult {
-        val scopes = discoveryRequest.context.getTests()
-            .map { testInfo ->
-                val matchingPath = discoveryRequest.paths.firstOrNull { it.intersects(testInfo.path) }
-                testInfo to matchingPath
-            }
-            .filter { (_, matchingPath) -> matchingPath != null }
-            .map { (testInfo, matchingPath) ->
-                checkNotNull(matchingPath)
-                val spec = resolveSpec(testInfo.createInstance(), testInfo.path)
-                spec.filterBy(matchingPath)
-                spec
+    private suspend fun CoroutineScope.filterScopes(discoveryRequest: DiscoveryRequest): Flow<GroupScopeImpl> = flow {
+        val tasks = mutableListOf<Deferred<GroupScopeImpl?>>()
+        discoveryRequest.context.getTests().forEach { testInfo ->
+            val matchingPath = discoveryRequest.paths.firstOrNull { it.intersects(testInfo.path) }
+            if (matchingPath != null) {
+                val task = async {
+                    val spec = resolveSpec(testInfo.createInstance(), testInfo.path)
+                    spec.filterBy(matchingPath)
+                    if (!spec.isEmpty()) {
+                        spec
+                    } else {
+                        null
+                    }
+
+                }
+                tasks.add(task)
             }
-            .filter { spec -> !spec.isEmpty() }
+        }
+
+        tasks.forEach { task ->
+            task.await()?.let { emit(it) }
+        }
+    }
 
+    @UseExperimental(ExperimentalTime::class)
+    fun discover(discoveryRequest: DiscoveryRequest): DiscoveryResult {
+        val scopes = mutableListOf<GroupScopeImpl>()
+        val time = measureTime {
+            doRunBlocking {
+                if (isConcurrentDiscoveryEnabled(false)) {
+                    withContext(Dispatchers.Default) {
+                        filterScopes(discoveryRequest).collect { scope ->
+                            scopes.add(scope)
+                        }
+                    }
+                } else {
+                    filterScopes(discoveryRequest).collect { scope ->
```

Comment:
Filtering in this path is done synchronously, since the dispatcher used is backed by the current thread only.